### ISSUE TITLE
`outerIIFEBody` option of `indent` for amd style definition

### ISSUE THREAD

**Issue by markelog** [eslint/eslint#6690]
**What version of ESLint are you using?**
3.1.0
**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

``` json
{
 "rules": { "indent": ["error", "tab", { "outerIIFEBody": 0 }] }
}
```

**What did you do? Please include the actual source code causing the issue.**

``` js
define( [
    "foo",
    "bar",
    "baz"
], function( foo, bar, baz ) {

test();

} );
```

**What did you expect to happen?**
Allow `test()` to not get indented

**What actually happened? Please include the actual, raw output from ESLint.**
It doesn't allow `test()` to not get indented

``` sh
7:1  error  Expected indentation of 1 tab character but found 0  indent
```

Needed for jQuery style see https://contribute.jquery.org/style-guide/js/#full-file-closures

**Comment by markelog** [comment]
@nzakas you fast with those labels, was about to add them :)

**Comment by nzakas** [comment]
I'm assuming what actually happened is that you got a warning about indentation and expected not to?

One thing to keep in mind: we don't provide anything that is specific to AMD in our rules, so we would need a suggestion for how to handle this in a generic way. 

Did JSCS handle this case?

**Comment by markelog** [comment]
> One thing to keep in mind: we don't provide anything that is specific to AMD in our rules, so we would need a suggestion for how to handle this in a generic way.

Yeah, I agree, it seems it could be generalized without specifically pointing to AMD, but i'm not sure how to correctly phrase it

> Did JSCS handle this case?

It was planned at some point, but nope.

I also will talk with the rest of the team, to just drop that rule, maybe we just peculiar that way

**Comment by markelog** [comment]
> with the rest of the team

I mean with jQuery team that is

**Comment by markelog** [comment]
> I'm assuming what actually happened is that you got a warning about indentation and expected not to?

Just updated ticket description, didn't expect to respond that fast

**Comment by nzakas** [comment]
Looking this over again, it looks like this should be a new option as we aren't actually taking about IIFEs in this case?

**Comment by TheCycoONE** [comment]
This is identical to the ```goog.scope``` case, where a particular function executes it's argument automatically. As far as I know, the only way to detect it would be to have a list of such functions.

**Comment by TheCycoONE** [comment]
It's worth noting that the Google style guide requires no indentation for the function defined within ```goog.scope```.

https://google.github.io/styleguide/javascriptguide.xml#goog-scope

**Comment by nzakas** [comment]
I think what we're really after here is anything in these formats:

```js
module.exports = function() {
// code
};

someFunction(function() {
// code
});
```

Where both `module.exports` and `someFunction` can be pretty much anything.

**Comment by TheCycoONE** [comment]
@nzakas were you thinking a user defined white list of functions to give special treatment to, or ambiguous indentation for top level functions?

**Comment by nzakas** [comment]
The latter. I think the simplest thing would just be to have an option that the first function expression that's either assigned to a global or passed as an argument to a global is not indented.

**Comment by jschaf** [comment]
I'm willing implement this as I need it for the `goog.scope` case.  Treating the first function expression as special works goog.scope.  As far as names for the option, how about.

- `firstFunctionExpressionBody` - follows same pattern as `outerIIFEBody`

**Comment by platinumazure** [comment]
Why the first? Seems to me like it should parallel `outerIIFEBody` and just be about "outermost" function expression assignments.

My suggestion would be to have two options:
* `outerFunctionExpressionAssignment` (covers `goog.scope` and `module.exports` and similar)
* `outerFunctionExpressionCallParameter` (covers `define()` and similar)

Open to suggestions on the names, of course.

**Comment by nzakas** [comment]
I'm fine with omitting the check for a function being first, but I'm not sure it makes sense to have two options. It seems one should suffice.

**Comment by ulitiy** [comment]
Agree with @platinumazure, we need to have a list of functions/function parameters, body of which are allowed to be unindented because there could be several of those in a single file. We can join them into one option somehow to avoid confusion.

**Comment by platinumazure** [comment]
@not-an-aardvark Just wondering, does the indent rewrite happen to include anything useful to this discussion at this point?

**Comment by not-an-aardvark** [comment]
No, it does not. It keeps the existing `outerIIFE` option, but doesn't add any new features related to this discussion.

**Comment by pschroen** [comment]
+1

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
